src/v/io/tests/cache_test.cc

```diff
@@ -0,0 +1,728 @@
+    // main entry is promoted, so the second is evicted. main queue size remains
+    // unchanged--it added one and evicted one.
+    EXPECT_FALSE(mains[0].evicted);
+    EXPECT_FALSE(mains[0].hook.evicted());
+    EXPECT_TRUE(mains[1].evicted);
+    EXPECT_TRUE(mains[1].hook.evicted());
+    EXPECT_EQ(cache->stat().main_queue_size, main_size);
+
+    for (auto& e : mains) {
+        cleanup(e);
+    }
+    for (auto& e : smalls) {
+        cleanup(e);
+    }
+}
+
+/*
+ * same as EvictSmallToFullMain except none of the entries in main are evictable
+ * so we'll fall back to evicting from the small queue.
+ */
+TEST_F(CacheTest, EvictSmallToFullMainPinned) {
+    // fill main queue to capacity
+    std::vector<entry> mains(main_size);
+    for (auto& e : mains) {
+        cache_insert_main(e);
+        e.may_evict = false;
+    }
+    EXPECT_EQ(cache->stat().main_queue_size, main_size);
+
+    // fill small queue right past capacity
+    std::vector<entry> smalls(small_size + 1);
+    for (auto& e : smalls) {
+        cache->insert(e);
+    }
+    EXPECT_EQ(cache->stat().small_queue_size, small_size + 1);
+
+    // the small queue is over capacity, so it will be evicted, but since it is
+    // a hot item we'll evict it to the main queue.
+    smalls[0].hook.touch();
+    smalls[0].hook.touch();
+    smalls[0].may_evict = false;
+    cache->evict();
+    EXPECT_FALSE(smalls[0].evicted);
+    EXPECT_FALSE(smalls[0].hook.evicted());
+
+    // but since main is at capacity, this will trigger main eviction. but none
+    // of the entries in main are evictable (including the small that was just
+    // promoted to main). so we'll fall back to small queue eviction.
+    EXPECT_EQ(cache->stat().small_queue_size, small_size - 1);
+    EXPECT_EQ(cache->stat().main_queue_size, main_size + 1);
+    for (const auto& e : mains) {
+        EXPECT_FALSE(e.evicted);
+        EXPECT_FALSE(e.hook.evicted());
+    }
+    EXPECT_TRUE(smalls[1].evicted);
+    EXPECT_TRUE(smalls[1].hook.evicted());
+
+    for (auto& e : mains) {
+        cleanup(e);
+    }
+    for (auto& e : smalls) {
+        cleanup(e);
+    }
+}
+
+TEST_F(CacheTest, NoEvictable) {
+    // fill main queue to capacity
+    std::vector<entry> mains(main_size);
+    for (auto& e : mains) {
+        cache_insert_main(e);
+    }
+
+    for (auto& e : mains) {
+        e.may_evict = false;
+    }
+
+    EXPECT_FALSE(cache->evict());
+
+    // fill small queue to capacity
+    std::vector<entry> smalls(small_size + 1);
+    for (auto& e : smalls) {
+        cache->insert(e);
+    }
+
+    for (auto& e : smalls) {
+        e.may_evict = false;
+    }
+
+    EXPECT_FALSE(cache->evict());
+
+    EXPECT_EQ(
+      cache->stat().small_queue_size + cache->stat().main_queue_size,
+      cache_size + 1);
+
+    for (auto& e : mains) {
+        cleanup(e);
+    }
+    for (auto& e : smalls) {
+        cleanup(e);
+    }
+}
+
+TEST(CacheTestCustom, CustomCost) {
```

Comment:
It would be great to add a test case that inserting while on the ghost queue inserts to the main queue and the small queue otherwise (or could you point me at it if i missed it?)

Reply:
I think this case is handled by `CacheTest.Insert` -- `cache_insert_main` helper simulates an entry being on the ghost queue.

Were you thinking about a different sort of test?

Reply:
Ah yeah that's true. I guess I was looking for a test that organically made its way to the ghost queue and then was inserted into the main queue. But it's true that cache_insert_main is already pretty much doing this, just with some tricks to get onto the ghost queue faster